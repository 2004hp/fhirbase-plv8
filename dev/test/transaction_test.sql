SET escape_string_warning=off;

\set cfg '{"base":"https://test.me"}'
\set alert `cat test/fixtures/alert.json`
\set bundle `cat test/fixtures/bundle.json`

BEGIN;

  WITH previous AS (
    SELECT
      fhir_create(:'cfg', 'Alert', :'alert'::jsonb, '[]'::jsonb)#>>'{entry,0,id}' AS update_id,
      fhir_create(:'cfg', 'Alert', :'alert'::jsonb, '[]'::jsonb)#>>'{entry,0,id}' AS delete_id
  ), bundle AS (
    SELECT
      p.*,
      replace(replace(:'bundle', '@previous-note', p.update_id), '@delete-note', p.delete_id) as bundle
    FROM previous p
  ), trans AS (
    SELECT
      b.update_id,
      b.delete_id,
      fhir_transaction(:'cfg', b.bundle::jsonb) as bundle
    FROM bundle b
  ), expanded AS (
    SELECT
      t.update_id,
      t.delete_id,
      t.bundle,
      t.bundle#>>'{entry,0,id}' as created_id,
      t.bundle#>>'{entry,1,id}' as updated_id,
      t.bundle#>>'{entry,2,id}' as deleted_id
    FROM trans t
  ), testing AS (
    SELECT
      e.*,
      fhir_read(:'cfg', 'Alert', e.created_id::uuid) as created,
      fhir_read(:'cfg', 'Alert', e.updated_id::uuid) as updated,
      fhir_read(:'cfg', 'Alert', e.deleted_id::uuid) as deleted
    FROM expanded e
  )

  SELECT
    assert_eq(
      t.update_id,
      t.updated_id,
      'update_id == updated_id'),
    assert_eq(
      t.delete_id,
      t.deleted_id,
      'delete_id & deleted_id'),
    assert_eq(
     t.bundle->>'totalResults',
      '3',
      'totalResults'),
    assert_eq(
      t.created#>>'{entry,0,content,note}',
      'create-note',
      'created note'),
    assert_eq(
      t.updated#>>'{entry,0,content,note}',
      'current-note',
      'updated note'),
    assert_eq(
      t.updated#>>'{entry,0,content,note}',
      'current-note',
      'updated note')
  FROM testing t;

ROLLBACK;
