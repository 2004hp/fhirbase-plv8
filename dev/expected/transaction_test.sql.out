SET escape_string_warning=off;
SET
BEGIN;
BEGIN
WITH previous AS (
    SELECT
      fhir_create('{"base":"https://test.me"}', 'Alert', '{
  "resourceType": "Alert",
   "status": "active",
   "subject": {
     "reference": "Patient/example",
     "display": "Peter Patient"
   },
   "note": "previous-note"
}'::jsonb, '[]'::jsonb)#>>'{entry,0,id}' AS update_id,
      fhir_create('{"base":"https://test.me"}', 'Device', '{
  "resourceType": "Device",
  "type": {
    "coding": [
      {
        "system": "http://snomed.info/sct",
        "code": "86184003",
        "display": "Electrocardiographic monitor and recorder"
      }
    ],
    "text": "ECG"
  }
}'::jsonb, '[]'::jsonb)#>>'{entry,0,id}' AS delete_id
  ), bundle AS (
    SELECT
      p.*,
      replace(replace('{
  "resourceType" : "Bundle",
  "entry" : [
    {
      "id" : "@create-device",
      "content" : {
        "resourceType": "Device",
        "type": {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "86184003",
              "display": "Electrocardiographic monitor and recorder"
            }
          ],
          "text": "ECG"
        }
      }
    },
    {
      "id" : "@update-note",
      "content" : {
        "resourceType": "Alert",
        "status": "active",
        "subject": {
          "reference": "@create-device",
          "display": "Peter Patient"
        },
        "note": "current-note",
        "author": {
          "reference": "@delete-device",
          "display": "Nancy Nurse"
        }
      }
    },
    {
      "id" : "@delete-device",
      "deleted" : "2012-05-29T23:45:32+00:00"
    }
  ]
}', '@update-note', p.update_id), '@delete-device', p.delete_id) as bundle
    FROM previous p
  ), trans AS (
    SELECT
      b.update_id,
      b.delete_id,
      fhir_transaction('{"base":"https://test.me"}', b.bundle::jsonb) as bundle
    FROM bundle b
  ), expanded AS (
    SELECT
      t.update_id,
      t.delete_id,
      t.bundle,
      t.bundle#>>'{entry,0,id}' as created_id,
      t.bundle#>>'{entry,1,id}' as updated_id,
      t.bundle#>>'{entry,2,id}' as deleted_id
    FROM trans t
  ), testing AS (
    SELECT
      e.*,
      fhir_read('{"base":"https://test.me"}', 'Device', e.created_id::uuid) as created,
      fhir_read('{"base":"https://test.me"}', 'Alert', e.updated_id::uuid) as updated,
      fhir_read('{"base":"https://test.me"}', 'Device', e.deleted_id::uuid) as deleted
    FROM expanded e
  )
  SELECT
    assert_eq(
      t.update_id,
      t.updated_id,
      'update_id == updated_id'),
    assert_eq(
      t.delete_id,
      t.deleted_id,
      'delete_id & deleted_id'),
    assert_eq(
     t.bundle->>'totalResults',
      '3',
      'totalResults'),
    assert_eq(
      t.created#>>'{entry,0,content,type,text}',
      'ECG',
      'created device'),
    assert_eq(
      t.updated#>>'{entry,0,content,note}',
      'current-note',
      'updated note'),
    assert_eq(
      t.deleted->>'entry',
      NULL,
      'updated device'),
    assert_eq(
      t.updated#>>'{entry,0,content,subject,reference}',
      t.created_id,
      'update created reference'),
    assert_eq(
      t.updated#>>'{entry,0,content,author,reference}',
      t.deleted_id,
      'update deleted reference')
  FROM testing t;
         assert_eq          |         assert_eq         |    assert_eq    |     assert_eq     |    assert_eq    |     assert_eq     |          assert_eq          |          assert_eq          
----------------------------+---------------------------+-----------------+-------------------+-----------------+-------------------+-----------------------------+-----------------------------
 OK update_id == updated_id | OK delete_id & deleted_id | OK totalResults | OK created device | OK updated note | OK updated device | OK update created reference | OK update deleted reference
(1 row)

ROLLBACK;
ROLLBACK
